tree/dataframe/inc/ROOT/RActionSnapshot.hxx

```diff
@@ -0,0 +1,257 @@
+#ifndef ROOT_RACTIONSNAPSHOT
+#define ROOT_RACTIONSNAPSHOT
+
+#include "ROOT/RDFNodes.hxx"
+#include "ROOT/RDFActionHelpers.hxx"
+#include "TError.h"
+
+namespace ROOT {
+namespace Internal {
+namespace RDF {
+
+class TypeErasedColumnValue {
+   std::shared_ptr<void> fPtr; // shared_ptr correctly deletes the type-erased object
+
+public:
+   template <typename T>
+   TypeErasedColumnValue(std::unique_ptr<RColumnValue<T>> v) : fPtr(std::move(v))
+   {
+   }
+
+   template <typename T>
+   T &Get(ULong64_t e)
+   {
+      return std::static_pointer_cast<RColumnValue<T>>(fPtr)->Get(e);
+   }
+
+   template <typename T>
+   RColumnValue<T> *Cast()
+   {
+      return static_cast<RColumnValue<T> *>(fPtr.get());
+   }
+};
+
+template <std::size_t... S, typename... ColTypes>
+void InitRDFValues(unsigned int slot, std::vector<TypeErasedColumnValue> &values, TTreeReader *r,
+                   const ColumnNames_t &bn, const RBookedCustomColumns &customCols, std::index_sequence<S...>,
+                   ROOT::TypeTraits::TypeList<ColTypes...>)
+{
+   std::array<bool, sizeof...(S)> isTmpColumn;
+   for (auto i = 0u; i < isTmpColumn.size(); ++i)
+      isTmpColumn[i] = customCols.HasName(bn.at(i));
+
+   using expander = int[];
+   (void)expander{(values.emplace_back(std::make_unique<RColumnValue<ColTypes>>()), 0)..., 0};
+   (void)expander{(isTmpColumn[S]
+                      ? values[S].Cast<ColTypes>()->SetTmpColumn(slot, customCols.GetColumns().at(bn.at(S)).get())
+                      : values[S].Cast<ColTypes>()->MakeProxy(r, bn.at(S)),
+                   0)...,
+                  0};
+}
+
+template <std::size_t... S, typename... ColTypes>
+void ResetRDFValueTuple(std::vector<TypeErasedColumnValue> &values, std::index_sequence<S...>,
+                        ROOT::TypeTraits::TypeList<ColTypes...>)
+{
+   using expander = int[];
+   (void)expander{(values[S].Cast<ColTypes>()->Reset(), 0)...};
+}
+
+template <typename PrevDataFrame, typename... ColTypes>
```

Comment:
the typed version is the generic `RAction<Helper>`, which we need to have better performance.
in the particular case of `RAction<SnapshotHelper>`, since the action itself goes to disk and is a "slow" operation, we don't necessarily need very fast access to the column values.